src/parser.cpp

```diff
@@ -135,16 +135,215 @@ namespace ada::parser {
     }
 
     // Convert ipv4 to string to use it inside "parse_host"
-    // TODO: Convert ipv4 to string and return
-    return std::make_tuple("", validation_error);
+    return std::make_tuple(std::to_string(ipv4), validation_error);
   }
 
   /**
    * @see https://url.spec.whatwg.org/#concept-ipv6-parser
    */
   parser_result<std::string_view> parse_ipv6(std::string_view input) {
-    // TODO: Implement this
-    return std::make_tuple(std::nullopt, false);
+    // Let address be a new IPv6 address whose IPv6 pieces are all 0.
+    std::vector<uint16_t> address(8);
```

Comment:
since this is fixed-size, std::array?

Reply:
that makes sense. is there a performance impact for this change?

Reply:
> that makes sense. is there a performance impact for this change?

It depends on the compiler and the optimization level, but with `std::vector`, you are much more likely to call `new` and get some heap allocation, and followed by a delete. A local std::array is almost surely 'on stack' so its memory usage comes at no performance cost. In theory, that might also happen with `std::vector<uint16_t> address(8)` and it will, some of the time.